In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
import logistic_regression_util
import acquire_walkthrough
import prepare_walkthrough


from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import sklearn.preprocessing


import warnings
warnings.filterwarnings("ignore")

import acquire
import prepare

In [2]:
df = acquire.get_titanic_data()

In [3]:
df

passenger_id  survived  pclass     sex   age  sibsp  parch     fare  \
0               0         0       3    male  22.0      1      0   7.2500   
1               1         1       1  female  38.0      1      0  71.2833   
2               2         1       3  female  26.0      0      0   7.9250   
3               3         1       1  female  35.0      1      0  53.1000   
4               4         0       3    male  35.0      0      0   8.0500   
..            ...       ...     ...     ...   ...    ...    ...      ...   
886           886         0       2    male  27.0      0      0  13.0000   
887           887         1       1  female  19.0      0      0  30.0000   
888           888         0       3  female   NaN      1      2  23.4500   
889           889         1       1    male  26.0      0      0  30.0000   
890           890         0       3    male  32.0      0      0   7.7500   

    embarked   class  deck  embark_town  alone  
0          S   Third  None  Southampton      0  
1          C   First     C    Cherbourg      0  
2          S   Third  None  Southampton      1  
3          S   First     C  Southampton      0  
4          S   Third  None  Southampton      1  
..       ...     ...   ...          ...    ...  
886        S  Second  None  Southampton      1  
887        S   First     B  Southampton      1  
888        S   Third  None  Southampton      0  
889        C   First     C    Cherbourg      1  
890        Q   Third  None   Queenstown      1  

[891 rows x 13 columns]

In [4]:
df.dropna(inplace=True)

In [5]:
df.head()

passenger_id  survived  pclass     sex   age  sibsp  parch     fare  \
1              1         1       1  female  38.0      1      0  71.2833   
3              3         1       1  female  35.0      1      0  53.1000   
6              6         0       1    male  54.0      0      0  51.8625   
10            10         1       3  female   4.0      1      1  16.7000   
11            11         1       1  female  58.0      0      0  26.5500   

   embarked  class deck  embark_town  alone  
1         C  First    C    Cherbourg      0  
3         S  First    C  Southampton      0  
6         S  First    E  Southampton      1  
10        S  Third    G  Southampton      0  
11        S  First    C  Southampton      1

In [6]:
df = df.rename(columns = {'sex':'gender'})

In [7]:
df.head(0)

Empty DataFrame
Columns: [passenger_id, survived, pclass, gender, age, sibsp, parch, fare, embarked, class, deck, embark_town, alone]
Index: []

In [8]:
test, train = prepare.prep_titanic(df)

In [9]:
train.head()

survived  pclass  gender       age  sibsp  parch      fare  embark_town  \
806         0       1    male  0.543379      0      0  0.000000  Southampton   
516         1       2  female  0.472032      0      0  0.020495  Southampton   
248         1       1    male  0.514840      1      1  0.102579  Southampton   
11          1       1  female  0.814498      0      0  0.051822  Southampton   
853         1       1  female  0.215183      0      1  0.076904  Southampton   

     alone  encoded_embark_town  
806      1                    2  
516      1                    2  
248      0                    2  
11       1                    2  
853      0                    2

In [10]:
X_train = train[['fare', 'pclass']]
y_train = train.survived
X_test = test[['fare', 'pclass']]
y_test = test.survived

In [11]:
logit = LogisticRegression(random_state = 123)

In [12]:
logit.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=123, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [13]:
y_pred = logit.predict(X_train)

In [14]:
model_df = train[['survived']]

In [15]:
model_df['yhat'] = logit.predict(X_train)
y = logit.predict_proba(X_train)[:, 1]
model_df['probs'] = y
model_df.head()

survived  yhat     probs
806         0     1  0.772498
516         1     1  0.783849
248         1     1  0.775639
11          1     1  0.774089
853         1     1  0.774856

In [16]:
model_df.head()

survived  yhat     probs
806         0     1  0.772498
516         1     1  0.783849
248         1     1  0.775639
11          1     1  0.774089
853         1     1  0.774856

In [17]:
print('Accuracy of Logistic Regression classifier on training set: {:.2f}'
     .format(logit.score(X_train, y_train)))

Accuracy of Logistic Regression classifier on training set: 0.81


In [18]:
X = train[['fare', 'pclass','age']]
y = train.survived

In [19]:
logit = LogisticRegression()

In [20]:
logit.fit(X, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [21]:
y_pred = logit.predict(X)

In [22]:
model_df['model2'] = logit.predict(X)

In [23]:
model_df.head()

survived  yhat     probs  model2
806         0     1  0.772498       1
516         1     1  0.783849       1
248         1     1  0.775639       1
11          1     1  0.774089       1
853         1     1  0.774856       1

In [24]:
# print('Accuracy of model 1 on training set: {:.2f}'
#      .format(logit.score(X_train, y_train)))
print('Accuracy of model 2 on training set: {:.2f}'
     .format(logit.score(X, y)))

Accuracy of model 2 on training set: 0.81


In [25]:
X1 = train[['fare', 'pclass','age','gender']]
y1 = train.survived

In [26]:
encoder = sklearn.preprocessing.OneHotEncoder()

encoder.fit(X1[['gender']])

OneHotEncoder(categorical_features=None, categories=None, drop=None,
              dtype=<class 'numpy.float64'>, handle_unknown='error',
              n_values=None, sparse=True)

In [27]:
m = encoder.transform(train[['gender']]).todense()

In [28]:
concat = pd.concat([
    X1.gender,
    pd.DataFrame(m, columns=encoder.categories_[0], index=X1.index)
], axis=1)

In [29]:
cols = ['gender_' + c for c in encoder.categories_[0]]

m = encoder.transform(X1[['gender']]).todense()
X1 = pd.concat([
    X1,
    pd.DataFrame(m, columns=cols, index=X1.index)
], axis=1).drop(columns='gender')

In [30]:
X1.head()

fare  pclass       age  gender_female  gender_male
806  0.000000       1  0.543379            0.0          1.0
516  0.020495       2  0.472032            1.0          0.0
248  0.102579       1  0.514840            0.0          1.0
11   0.051822       1  0.814498            1.0          0.0
853  0.076904       1  0.215183            1.0          0.0

In [31]:
logit.fit(X1, y1)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [32]:
y_pred = logit.predict(X1)

In [33]:
model_df['model3'] = logit.predict(X1)

In [34]:
model_df.head()

survived  yhat     probs  model2  model3
806         0     1  0.772498       1       1
516         1     1  0.783849       1       1
248         1     1  0.775639       1       1
11          1     1  0.774089       1       1
853         1     1  0.774856       1       1

In [35]:
print('Accuracy of model 3 on training set: {:.2f}'
     .format(logit.score(X1, y1)))

Accuracy of model 3 on training set: 0.81


In [36]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_graphviz
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [37]:
titanic = acquire_walkthrough.get_titanic_data()

In [38]:
def encode_sex(df):
    '''
    Returns a new dataframe with the ``sex`` column encoded.
    '''
    return df.assign(
        sex=(df.sex == 'female').astype(int)
    )

In [39]:
def get_splits(titanic):
    '''
    Returns X and y for train, validate and test datasets
    '''
    # don't blow away our original data
    titanic = titanic.copy()
    
    # ignore warnings just for this block
    with warnings.catch_warnings():
        warnings.simplefilter('ignore')
        scaler, encoder, train, test = prepare_walkthrough.prep_titanic(titanic)
    
    # Which features are we going to look at?
    cols = ['survived', 'pclass', 'sex', 'age', 'alone']
    train = train[cols]
    test = test[cols]

    # validate data split
    train, validate = sklearn.model_selection.train_test_split(
        train, train_size=.85, random_state=123
    )

    # split into X and y
    X_train, y_train = train.drop(columns='survived'), train[['survived']]
    X_validate, y_validate = validate.drop(columns='survived'), validate[['survived']]
    X_test, y_test = test.drop(columns='survived'), test[['survived']]
    
    X_train = encode_sex(X_train)
    X_validate = encode_sex(X_validate)
    X_test = encode_sex(X_test)    
    
    return X_train, y_train, X_validate, y_validate, X_test, y_test

In [40]:
X_train, y_train, X_validate, y_validate, X_test, y_test = get_splits(titanic)

print('   train: %d rows' % X_train.shape[0])
print('validate: %d rows' % X_validate.shape[0])
print('    test: %d rows' % X_test.shape[0])

   train: 567 rows
validate: 101 rows
    test: 223 rows


In [42]:
# a dataframe to hold our models' predictions for future comparison
evaluation = pd.DataFrame({
    'actual': y_validate.survived
})

In [43]:
evaluation.head()

actual
371       0
57        0
398       0
376       1
194       1

In [44]:
clf = DecisionTreeClassifier(criterion='entropy', random_state=123)

In [45]:
clf.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=123, splitter='best')

In [46]:
y_pred = clf.predict(X_train)
y_pred_proba = clf.predict_proba(X_train)

In [50]:
conf = pd.DataFrame(confusion_matrix(y_train, y_pred))
conf.index.name = "actual"
conf

0    1
actual          
0       344    9
1        44  170

In [55]:
labels = ['dead', 'survived']
predicted_labels = [name + " predicted" for name in labels ]

conf = pd.DataFrame(confusion_matrix(y_pred, y_train), index=labels, columns=[predicted_labels])
conf.index.name = "actual"
conf

dead predicted survived predicted
actual                                    
dead                344                 44
survived              9                170

In [52]:
print('Accuracy of Decision Tree classifier on training set: {:.2f}'
     .format(clf.score(X_train, y_train)))

Accuracy of Decision Tree classifier on training set: 0.91


In [53]:
pd.DataFrame(classification_report(y_train, y_pred, output_dict=True))

0           1  accuracy   macro avg  weighted avg
precision    0.886598    0.949721  0.906526    0.918159      0.910422
recall       0.974504    0.794393  0.906526    0.884448      0.906526
f1-score     0.928475    0.865140  0.906526    0.896807      0.904571
support    353.000000  214.000000  0.906526  567.000000    567.000000

- TP: predicted survived, actually survived (170)
- FP: predicted survived, acutally ~survived (9)
- FN: predicted ~survived, acutally survived (44)
- TN: predicted ~survived, actually ~survived (344)

- __accuracy__: how well the model does overall
> __(TP + TN) / (TP + TN + FP + FN)__
(170 + 344) / (170 + 9 + 44 + 344) = 90.6%
- __precision__: % of actually positive cases that were predicted as positive
> __TP / (TP + FP)__
170 / (170 + 9) = 94.9%
- __recall__: % of of positive predictions that are correct
> __TP / (TP + FN)__
170 / (170 + 44) = 79.4%

In [65]:
y_pred = clf.predict(X_validate)
y_pred_proba = clf.predict_proba(X_validate)

In [66]:
labels = ['dead', 'survived']
predicted_labels = [name + " predicted" for name in labels ]

conf = pd.DataFrame(confusion_matrix(y_pred, y_validate), index=labels, columns=[predicted_labels])
conf.index.name = "actual"
conf

dead predicted survived predicted
actual                                    
dead                 56                 11
survived              3                 31

In [67]:
print('Accuracy of Decision Tree classifier on training set: {:.2f}'
     .format(clf.score(X_validate, y_validate)))

Accuracy of Decision Tree classifier on training set: 0.86


In [68]:
pd.DataFrame(classification_report(y_validate, y_pred, output_dict=True))

0          1  accuracy   macro avg  weighted avg
precision   0.835821   0.911765  0.861386    0.873793      0.867401
recall      0.949153   0.738095  0.861386    0.843624      0.861386
f1-score    0.888889   0.815789  0.861386    0.852339      0.858491
support    59.000000  42.000000  0.861386  101.000000    101.000000